plugins/pixart-rf/fu-pxi-device.c

```diff
@@ -0,0 +1,611 @@
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#ifdef HAVE_HIDRAW_H
+#include <linux/hidraw.h>
+#include <linux/input.h>
+#endif
+
+#include "fu-common.h"
+#include "fu-chunk.h"
+
+#include "fu-pxi-device.h"
+#include "fu-pxi-firmware.h"
+
+#define PXI_HID_DEV_OTA_INPUT_REPORT_ID		0x05
+#define PXI_HID_DEV_OTA_OUTPUT_REPORT_ID	0x06
+#define PXI_HID_DEV_OTA_FEATURE_REPORT_ID	0x07
+
+#define FU_PXI_DEVICE_CMD_FW_OTA_INIT		0x10u
+#define FU_PXI_DEVICE_CMD_FW_WRITE		0x17u
+#define FU_PXI_DEVICE_CMD_FW_UPGRADE		0x18u
+#define FU_PXI_DEVICE_CMD_FW_MCU_RESET		0x22u	/* remove after windows OTA app ready */
+#define FU_PXI_DEVICE_CMD_FW_GET_INFO		0x23u
+#define FU_PXI_DEVICE_CMD_FW_OBJECT_CREATE	0x25u
+#define FU_PXI_DEVICE_CMD_FW_OTA_INIT_NEW	0x27u
+#define FU_PXI_DEVICE_CMD_FW_OTA_RETRANSMIT	0x28u
+#define FU_PXI_DEVICE_CMD_FW_OTA_DISCONNECT	0x29u
+
+#define FU_PXI_DEVICE_OBJECT_SIZE_MAX		4096	/* bytes */
+#define FU_PXI_DEVICE_OTA_BUF_SZ		32	/* bytes */
+#define FU_PXI_DEVICE_NOTTFY_RET_LEN		4	/* bytes */
+#define FU_PXI_DEVICE_FW_INFO_RET_LEN		8	/* bytes */
+
+/* OTA target selection */
+enum ota_process_setting {
+	OTA_MAIN_FW,				/* Main firmware */
+	OTA_HELPER_FW,				/* Helper firmware */
+	OTA_EXTERNAL_RESOURCE,			/* External resource */
+};
+
+/* OTA spec check result */
+enum ota_spec_check_result {
+	OTA_SPEC_CHECK_OK		= 1,	/* Spec check ok */
+	OTA_FW_OUT_OF_BOUNDS		= 2,	/* OTA firmware size out of bound */
+	OTA_PROCESS_ILLEGAL		= 3,	/* Illegal OTA process */
+	OTA_RECONNECT			= 4,	/* Inform OTA app do reconnect */
+	OTA_FW_IMG_VERSION_ERROR	= 5,	/* FW image file version check error */
+	OTA_SPEC_CHECK_MAX_NUM,			/* Max number of OTA driver defined error code */
+};
+
+/* OTA disconnect reason */
+enum ota_disconnect_reason {
+	OTA_CODE_JUMP			= 1,	/* OTA code jump */
+	OTA_UPDATE_DONE			= 2,	/* OTA update done */
+	OTA_RESET,				/* OTA reset */
+};
+
+struct _FuPxiDevice {
+	FuUdevDevice	 parent_instance;
+	guint8		 status;
+	guint8		 new_flow;
+	guint16		 offset;
+	guint16		 checksum;
+	guint32		 max_object_size;
+	guint16		 mtu_size;
+	guint16		 prn_threshold;
+	guint8		 spec_check_result;
+};
+
+G_DEFINE_TYPE (FuPxiDevice, fu_pxi_device, FU_TYPE_UDEV_DEVICE)
+
+static void
+fu_pxi_device_to_string (FuDevice *device, guint idt, GString *str)
+{
+	FuPxiDevice *self = FU_PXI_DEVICE (device);
+	fu_common_string_append_kx (str, idt, "Status", self->status);
+	fu_common_string_append_kx (str, idt, "NewFlow", self->new_flow);
+	fu_common_string_append_kx (str, idt, "CurrentObjectOffset", self->offset);
+	fu_common_string_append_kx (str, idt, "CurrentChecksum", self->checksum);
+	fu_common_string_append_kx (str, idt, "MaxObjectSize", self->max_object_size);
+	fu_common_string_append_kx (str, idt, "MtuSize", self->mtu_size);
+	fu_common_string_append_kx (str, idt, "PacketReceiptNotificationThreshold", self->prn_threshold);
+	fu_common_string_append_kx (str, idt, "SpecCheckResult", self->spec_check_result);
```

Comment:
given there is actually an enum that decodes this, I think it makes sense to have a to_string helper for it that gives you messages.

Reply:
Agree, fixed.